doc/man/diffractometeralignment.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.4.
.TH DIFFRACTOMETERALIGNMENT "1" "November 2019" "diffractometeralignment 2.8.6" "User Commands"
.SH NAME
diffractometeralignment \- manual page for diffractometeralignment 2.8.6
.SH SYNOPSIS
.B diffractometeralignment
\fI\,<model> \/\fR[\fI\,door_name\/\fR]
.SH DESCRIPTION
a taurus application for diffractometer alignment: h, k, l movements and
scans, go to maximum, ...
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-\-version\fR
show program's version number and exit
.IP
Taurus Options:
.IP
Basic options present in any taurus application
.TP
\fB\-\-taurus\-log\-level\fR=\fI\,LEVEL\/\fR
taurus log level. Allowed values are (case
insensitive): critical, error, warning/warn, info,
debug, trace
.TP
\fB\-\-taurus\-polling\-period\fR=\fI\,MILLISEC\/\fR
taurus global polling period in milliseconds
.TP
\fB\-\-taurus\-serialization\-mode\fR=\fI\,SERIAL\/\fR
taurus serialization mode. Allowed values are (case
insensitive): serial, concurrent (default)
.TP
\fB\-\-tango\-host\fR=\fI\,TANGO_HOST\/\fR
Tango host name (either HOST:PORT or a Taurus URI,
e.g. tango://foo:1234)
.TP
\fB\-\-remote\-console\-port\fR=\fI\,PORT\/\fR
enables remote debugging using the given port
.TP
\fB\-\-default\-formatter\fR=\fI\,FORMATTER\/\fR
Override the default formatter


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).
This file follows the formats and conventions from [keepachangelog.com]

## [Unreleased]

### Added

* Support to Python >= 3.5 (#1089, #1173, #1201, #1313, #1336)
* Showscan online based on pyqtgraph (#1285)
  * multiple plots in the same MultiPlot widget (as opposed to different panels before)
  * option to group curves by x-axis or individual plot per curve
  * new showscan console script
  * support fast scans: update curves at a fix rate (5Hz)
  * better curve colors and symbols
* Measurement group (Taurus extension) configuration API with methods to
  set/get: enabled, output, plot type, plot axes, timer, monitor, synchronizer,
  value ref enabled, value ref pattern parameters(#867)
* Experiment configuration (expconf) macros
  * Measurement group configuration macros: `set_meas_conf` and `get_meas_conf` (#690)
  * Active measurement group selection macros: `set_meas` and `get_meas` (#690)
  * Pre-scan snapshot macros: `lssnap`, `defsnap` and `udefsnap` (#1199)
* Automatic scan statistics calculation with the `scanstats` macro as the `post-scan`
  hook stored in the `ScanStats` environment variable (#880)
* `pic`, `cen` to move the scanned motor to the peak and center of FWHM values
  respectively (#890)
* `where` macro to print the scanned motor position (#890)
* `plotselect` macro for configuring channels for online scan plotting (#824)
* `genv` macro for printing environment variable values (#888)
* Dump info on channels if MG acq fails in step scan, ct and uct (#1308)
* Add timestamp to element's dumped information (#1308)
* Quality to `SardanaAttribute` (#1353)
* Instruments creation and configuration in sar_demo (#1198)
* Allow _experimental channel acquisition_ with PoolChannelTaurusValue (PCTV) widget (#1203)
* Documentation to Taurus Extensions of Sardana Devices: MacroServer part
  and the whole Sardana part of the Qt Taurus Extensions (#1228, #1233)
* Advertise newfile macro in case no ScanDir or ScanFile is set (#1254, #1258)
* Improve scans to detect if a ScanFile od ScanDir are set but empty (#1262)
* Possibility to view debug messages in `DoorOutput` widget - enable/disable
  using context menu option (#1242)
* Improve user experience with PMTV:
  * Store PMTV (motor widget) configurations: *expert view* and *write mode*
    (relative or absolute) permanently as TaurusGUI settings (#1286)
  * Do not create encoder widget in PMTV if the motod does not have encoder
    in order to avoid errors comming from the polling (#209, #1288)
  * Change limit switches indicators from buttons to labels (#210, #1290)
* Improve documentation (#1241)
* Better macro exception message and hint to use `www` (#1191)
* Stress tests (on the Taurus level) for measurements (#1353)
* Add basic information to "how to write custom recorder" to
  the documentation (#1275)
* Register a TaurusValue factory for pool widgets (#1333)
* Direct links to Sardana-Taurus model API (#1335)

### Fixed

* Use `tango.EnsureOmnitThread` to protect Sardana threads
  (Tango is not thread safe) (#1298)
* Avoid using Tango `AttributeProxy` in limits protection to not be affected
  by bug tango-controls/pytango#315 (#1302)
* Avoid deadlock in Sardana-Taurus models e.g. `MeasurementGroup.count()` or
  `Motor.move()` (#1348)
  * Remove redundant protection in PoolElement.start() and waitFinish()
* Fast repetitions of single acqusition measurements (counts) on MeasurementGroup (#1353)
* Pre-mature returning to ON state of MeasurementGroup at the end of measurement (#1353) 
* Default macro parameter values in macroexecutor (#1153)
* Executing RunMacro Door's command with string parameters containing spaces (#1240)
* Setting of environment variables in Python 3.7 (#1195)
* Use `taurus.external.qt.compat.PY_OBJECT` in singal signatures instead of `object`
  to avoid problems when using `builtins` from `future` (#1082)
* Remove Taurus deprecated code what reduces deprecation warnings (#1206, #1252)
* Macro functions which define results now really report the result (#1238)
* Use of env and hints in `macro` function decorator (#1239)
* Fix several issues with PMTV:
  * Reset pending operations of absolute movement on switching to relative movement (#1293)
  * PMTV widget not updating the following attributes: limit switches, state
    and status (#1244)
* Avoid Taurus GUI slowness on startup and changing of perspectives due to too
  large macroexecutor history by limitting it to 100 -
  configurable with `MACROEXECUTOR_MAX_HISTORY` (#1307)
* OutputBlock view option when macros produce outputs at high rate (#1245)
* `showscan online` shows only the online trend and not erroneously online and offline
  (#1260)
* Fix fast operations (motion & acq) by propertly clearing operation context and
  resetting of acq ctrls dicts (#1300)
* Use more efficient way to get terminal size for better printing spock output (#1245, #1268)
* Measurement groups renaming with `renameelem` macro(#951)
* `macroexecutor` correctly loads macro combo box if it was started with server down and 
  server started afterwards (#599, #1278)
* `TaurusMacroExecutorWidget` does not use _parent model_ feature (#599, #1278)
* `TaurusSequencerWidget` does not use _parent model_ feature (#1284)
* Macro plotting in new versions of ipython and matplotlib require extra call to
  `pyplot.draw()` to make sure that the plot is refreshed (#1280)
* Allow MacroButton widget to be smaller - minimum size to show the macro name (#1265)
* Remove TangoAttribute controllers from Sardana (#181, #1279)
* Remove deprecation warning revealed when running test suite (#1267)
* Remove event filtering in `DynamicPlotManager` (showscan online) (#1299)
* Avoid unnecessary creations of DeviceProxies in `ascanct` (#1281)
* Adapt to new taurus behavior of `cmd_line_parser` kwarg of `TaurusApplication` (#1306)

src/sardana/taurus/qt/qtgui/extra_hkl/diffractometeralignment.py
#!/usr/bin/env python

##############################################################################
##
# This file is part of Sardana
##
# http://www.sardana-controls.org/
##
# Copyright 2011 CELLS / ALBA Synchrotron, Bellaterra, Spain
##
# Sardana is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
##
# Sardana is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
##
# You should have received a copy of the GNU Lesser General Public License
# along with Sardana.  If not, see <http://www.gnu.org/licenses/>.
##
##############################################################################

__docformat__ = 'restructuredtext'

import sys
import time
from taurus.external.qt import Qt
# TODO: avoid using of PyTango - use Taurus instead
import PyTango
import sardana

from taurus.qt.qtgui.container import TaurusWidget
from taurus.qt.qtgui.display import TaurusLabel
from taurus.qt.qtgui.base import TaurusBaseWidget

from taurus.external.qt import QtCore, QtGui

from taurus.qt.qtcore.communication import SharedDataManager
from taurus.qt.qtgui.input import TaurusValueLineEdit

import taurus.core.util.argparse
import taurus.qt.qtgui.application
from taurus.qt.qtgui.util.ui import UILoadable

from sardana.taurus.qt.qtgui.extra_macroexecutor import TaurusMacroConfigurationDialog

from .selectsignal import SelectSignal


class EngineModesComboBox(Qt.QComboBox, TaurusBaseWidget):
    """ComboBox representing list of engine modes"""

    def __init__(self, parent=None):
        name = self.__class__.__name__
        self.call__init__wo_kw(Qt.QComboBox, parent)
        self.call__init__(TaurusBaseWidget, name)
        self.setSizeAdjustPolicy(Qt.QComboBox.AdjustToContentsOnFirstShow)
        self.setToolTip("Choose a engine mode ...")
        QtCore.QMetaObject.connectSlotsByName(self)

    def loadEngineModeNames(self, enginemodes):
        self.clear()
        self.addItems(enginemodes)


@UILoadable(with_ui="_ui")
class DiffractometerAlignment(TaurusWidget):

    def __init__(self, parent=None, designMode=False):
        TaurusWidget.__init__(self, parent, designMode=designMode)

        self.loadUi(filename="diffractometeralignment.ui")

        self.selectsignal = SelectSignal()

        self._ui.AlignmentStopButton.clicked.connect(self.stop_movements)
        self._ui.AlignmentStoreReflectionButton.clicked.connect(
            self.store_reflection)

        self._ui.MacroServerConnectionButton.clicked.connect(
            self.open_macroserver_connection_panel)

        self._ui.SelectSignalButton.clicked.connect(
            self.open_selectsignal_panel)

        # Create a global SharedDataManager
        Qt.qApp.SDM = SharedDataManager(self)

    @classmethod
    def getQtDesignerPluginInfo(cls):
        ret = TaurusWidget.getQtDesignerPluginInfo()
        ret['module'] = 'diffractometeralignment'
        ret['group'] = 'Taurus Containers'
        ret['container'] = ':/designer/frame.png'
        ret['container'] = False
        return ret

    def setModel(self, model):
        if model is not None:
            self.device = taurus.Device(model)

        self.pseudo_motor_names = []
        for motor in self.device.hklpseudomotorlist:
            self.pseudo_motor_names.append(motor.split(' ')[0])

        self.h_device_name = self.pseudo_motor_names[0]
        self.h_device = taurus.Device(self.h_device_name)
        self.k_device_name = self.pseudo_motor_names[1]
        self.k_device = taurus.Device(self.k_device_name)
        self.l_device_name = self.pseudo_motor_names[2]
        self.l_device = taurus.Device(self.l_device_name)

        # Set model to hkl components

        hmodel = self.h_device_name + "/Position"
        self._ui.taurusValueLineH.setModel(hmodel)
        self._ui.taurusLabelValueH.setModel(hmodel)
        kmodel = self.k_device_name + "/Position"
        self._ui.taurusValueLineK.setModel(kmodel)
        self._ui.taurusLabelValueK.setModel(kmodel)
        lmodel = self.l_device_name + "/Position"
        self._ui.taurusValueLineL.setModel(lmodel)
        self._ui.taurusLabelValueL.setModel(lmodel)

        # Add dynamically the angle widgets

        motor_list = self.device.motorlist
        self.motor_names = []
        self.motor_devices = []

        for motor in self.device.motorlist:
            self.motor_names.append(motor.split(' ')[0])
            self.motor_devices.append(taurus.Device(
                self.motor_names[len(self.motor_names) - 1]))

        self.nb_motors = len(motor_list)

        angles_labels = []
        self.angles_names = []
        angles_taurus_label = []
        angles_taurus_input = []

        gap_x = 650 // self.nb_motors

        try:
            self.angles_names = self.device.motorroles
        except:  # Only for compatibility
            if self.nb_motors == 4:
                self.angles_names.append("omega")
                self.angles_names.append("chi")
                self.angles_names.append("phi")
                self.angles_names.append("theta")
            elif self.nb_motors == 6:
                self.angles_names.append("mu")
                self.angles_names.append("th")
                self.angles_names.append("chi")
                self.angles_names.append("phi")
                self.angles_names.append("gamma")
                self.angles_names.append("delta")

        for i in range(0, self.nb_motors):
            angles_labels.append(QtGui.QLabel(self))
            angles_labels[i].setGeometry(
                QtCore.QRect(150 + gap_x * i, 40, 71, 17))
            angles_labels[i].setLayoutDirection(QtCore.Qt.RightToLeft)
            alname = "angleslabel" + str(i)
            angles_labels[i].setObjectName(alname)
            angles_labels[i].setText(QtGui.QApplication.translate(
                "HKLScan", self.angles_names[i], None))

            angles_taurus_label.append(TaurusLabel(self))
            angles_taurus_label[i].setGeometry(
                QtCore.QRect(150 + gap_x * i, 70, 81, 19))
            atlname = "anglestauruslabel" + str(i)
            angles_taurus_label[i].setObjectName(atlname)
            angles_taurus_label[i].setModel(self.motor_names[i] + "/Position")

            angles_taurus_input.append(TaurusValueLineEdit(self))
            angles_taurus_input[i].setGeometry(
                QtCore.QRect(145 + gap_x * i, 100, 91, 27))
            atlname = "anglestaurusinput" + str(i)
            angles_taurus_input[i].setObjectName(atlname)
            angles_taurus_input[i].setModel(self.motor_names[i] + "/Position")

        # Set model to engine and modes

        enginemodel = model + '/engine'
        self._ui.taurusLabelEngine.setModel(enginemodel)
        enginemodemodel = model + '/enginemode'
        self._ui.taurusLabelEngineMode.setModel(enginemodemodel)

        self.enginemodescombobox = EngineModesComboBox(self)
        self.enginemodescombobox.setGeometry(QtCore.QRect(150, 315, 221, 27))
        self.enginemodescombobox.setObjectName("enginemodeslist")

        self.enginemodescombobox.loadEngineModeNames(self.device.hklmodelist)

        self.enginemodescombobox.currentIndexChanged['QString'].connect(
            self.onModeChanged)

        # Add dynamically the scan buttons, range inputs and 'to max' buttons

        scan_buttons = []
        self.range_inputs = []
        self.tomax_buttons = []  # The text will be change when the max. is computed

        exec_functions = [self.exec_scan1, self.exec_scan2, self.exec_scan3,
                          self.exec_scan4, self.exec_scan5, self.exec_scan6]

        tomax_functions = [self.tomax_scan1, self.tomax_scan2, self.tomax_scan3,
                           self.tomax_scan4, self.tomax_scan5, self.tomax_scan6]

        gap_x = 650 // self.nb_motors

        for i in range(0, self.nb_motors):
            scan_buttons.append(QtGui.QPushButton(self))
            scan_buttons[i].setGeometry(
                QtCore.QRect(150 + gap_x * i, 405, 100, 26))
            wname = "scanbutton" + str(i)
            scan_buttons[i].setObjectName(wname)
            scan_buttons[i].setText(QtGui.QApplication.translate(
                "DiffractometerAlignment", self.angles_names[i], None))
            scan_buttons[i].clicked.connect(exec_functions[i])

            self.range_inputs.append(QtGui.QLineEdit(self))
            self.range_inputs[i].setGeometry(
                QtCore.QRect(150 + gap_x * i, 440, 100, 26))
            self.range_inputs[i].setLayoutDirection(QtCore.Qt.RightToLeft)
            wname = "rangeinput" + str(i)
            self.range_inputs[i].setObjectName(wname)

            self.tomax_buttons.append(QtGui.QPushButton(self))
            self.tomax_buttons[i].setGeometry(
                QtCore.QRect(150 + gap_x * i, 475, 100, 26))
            wname = "tomaxbutton" + str(i)
            self.tomax_buttons[i].setObjectName(wname)
            self.tomax_buttons[i].setText(QtGui.QApplication.translate(
                "DiffractometerAlignment", 'n.n.', None))
            self.tomax_buttons[i].clicked.connect(tomax_functions[i])

    def exec_scan1(self):
        self.exec_scan(0)

    def exec_scan2(self):
        self.exec_scan(1)

    def exec_scan3(self):
        self.exec_scan(2)

    def exec_scan4(self):
        self.exec_scan(3)

    def exec_scan5(self):
        self.exec_scan(4)

    def exec_scan6(self):
        self.exec_scan(5)

    def exec_scan(self, imot):
        macro_command = []

        macro_command.append("_diff_scan")
        macro_command.append(str(self.motor_names[imot]))
        current_pos = self.motor_devices[imot].Position
        range_scan = float(self.range_inputs[imot].text())
        macro_command.append(str(current_pos - range_scan))
        macro_command.append(str(current_pos + range_scan))
        macro_command.append(str(self._ui.NbPointslineEdit.text()))
        macro_command.append(
            str(self.selectsignal._ui.SampleTimelineEdit.text()))
        macro_command.append(str(self.selectsignal._ui.SignallineEdit.text()))

        self.door_device.RunMacro(macro_command)
        while(self.door_device.State()) == PyTango.DevState.RUNNING:
            time.sleep(0.01)
        # TODO: the string parsing should be eliminated and the sardana
        # generic "goto_peak" feature should be used instead - when available
        output_values = self.door_device.read_attribute("Output").value
        if output_values is not None:
            for i in range(len(output_values)):
                if output_values[i] == "Position to move":
                    self.tomax_buttons[imot].setText(QtGui.QApplication.translate(
                        "DiffractometerAlignment", str(output_values[i + 1]),
                        None))

    def tomax_scan1(self):
        self.tomax_scan(0)

    def tomax_scan2(self):
        self.tomax_scan(1)

    def tomax_scan3(self):
        self.tomax_scan(2)

    def tomax_scan4(self):
        self.tomax_scan(3)

    def tomax_scan5(self):
        self.tomax_scan(4)

    def tomax_scan6(self):
        self.tomax_scan(5)

    def tomax_scan(self, imot):
        motor = str(self.motor_names[imot])
        position = str(self.tomax_buttons[imot].text())
        macro_command = ["mv", motor, position]
        self.door_device.RunMacro(macro_command)

    @Qt.pyqtSlot('QString')
    def onModeChanged(self, modename):
        if self.device.engine != "hkl":
            self.device.write_attribute("engine", "hkl")
        self.device.write_attribute("enginemode", str(modename))

    def open_macroserver_connection_panel(self):
        w = TaurusMacroConfigurationDialog(self)
        Qt.qApp.SDM.connectReader("macroserverName", w.selectMacroServer)
        Qt.qApp.SDM.connectReader("doorName", w.selectDoor)
        Qt.qApp.SDM.connectReader("doorName", self.onDoorChanged)
        Qt.qApp.SDM.connectWriter(
            "macroserverName", w, 'macroserverNameChanged')
        Qt.qApp.SDM.connectWriter("doorName", w, 'doorNameChanged')

        w.show()

    def onDoorChanged(self, doorName):
        if doorName != self.door_device_name:
            self.door_device_name = doorName
            self.door_device = taurus.Device(doorName)

    def stop_movements(self):
        self.door_device.StopMacro()

    def store_reflection(self):
        hklref = []
        hklref.append(self.h_device.Position)
        hklref.append(self.k_device.Position)
        hklref.append(self.l_device.Position)

        self.device.write_attribute("addreflection", hklref)

    def open_selectsignal_panel(self):

        self.selectsignal.update_signals(self.door_device_name)
        self.selectsignal.show()


def main():
    parser = taurus.core.util.argparse.get_taurus_parser()
    parser.usage = "%prog <model> [door_name]"
    desc = ("a taurus application for diffractometer alignment: h, k, l " +
            "movements and scans, go to maximum, ...")
    parser.set_description(desc)

    app = taurus.qt.qtgui.application.TaurusApplication(
        cmd_line_parser=parser, app_version=sardana.Release.version)
    app.setApplicationName("diffractometeralignment")
    args = app.get_command_line_args()
    if len(args) < 1:
        msg = "model not set (requires diffractometer controller)"
        parser.error(msg)

    w = DiffractometerAlignment()
    w.model = args[0]
    w.setModel(w.model)

    w.door_device = None
    w.door_device_name = None
    if len(args) > 1:
        w.onDoorChanged(args[1])
    else:
        msg = ("No door name supplied. Connection to MacroServer/Door will " +
               "not not automatically done.")
        app.warning(msg)

    w.show()

    sys.exit(app.exec_())


if __name__ == "__main__":
    main()
